kj-rs/own.rs

```diff
@@ -0,0 +1,183 @@
+//! The `workerd-cxx` module containing the [`KjOwn<T>`] type, which is bindings to the `kj::Own<T>` C++ type
+use crate::fmt::display;
+use std::ffi::c_void;
+use std::fmt::{self, Debug, Display};
+use std::marker::PhantomData;
+use std::mem::MaybeUninit;
+use std::ops::Deref;
+use std::ops::DerefMut;
+use std::pin::Pin;
+
+/// A [`KjOwn<T>`] represents the `kj::Own<T>`. It is a smart pointer to an opaque C++ type.
+#[repr(C)]
+pub struct KjOwn<T>
+where
+    T: KjOwnTarget,
+{
+    repr: [MaybeUninit<*mut c_void>; 2],
+    _ty: PhantomData<T>,
+}
+
+// Possible Other Types:
+// - SpaceFor<T>/construct>
+
+// Public-facing KjOwn api, backed by calls to unsafe code generated for each [`KjOwnTarget`]
+impl<T> KjOwn<T>
+where
+    T: KjOwnTarget,
+{
+    // Possible functions can include:
+    // [`KjOwn::heap`] for allocating C++ types from rust, by calling into the C++ `kj::heap` function
+    // [`KjOwn::fakeOwn`]
+    // [`KjOwn::refCounted`]
+    // [`KjOwn::attachRef`]
+
+    /// Returns a mutable pinned reference to the object owned by this [`KjOwn`]
+    /// if any, otherwise None.
+    pub fn as_mut(&mut self) -> Option<Pin<&mut T>> {
+        let this = std::ptr::from_ref::<Self>(self).cast::<c_void>();
+        unsafe {
+            let mut_reference = T::__get(this).cast_mut().as_mut()?;
+            Some(Pin::new_unchecked(mut_reference))
+        }
+    }
+
+    /// Returns a reference to the object owned by this [`KjOwn`] if any,
+    /// otherwise None.
+    #[must_use]
+    pub fn as_ref(&self) -> Option<&T> {
+        let this = std::ptr::from_ref::<Self>(self).cast::<c_void>();
+        unsafe { T::__get(this).as_ref() }
+    }
+
+    /// Returns a mutable pinned reference to the object owned by this
+    /// [`KjOwn`].
+    ///
+    /// # Panics
+    ///
+    /// Panics if the [`KjOwn`] holds a null pointer.
+    pub fn pin_mut(&mut self) -> Pin<&mut T> {
+        match self.as_mut() {
+            Some(target) => target,
+            None => panic!("called pin_mut on a null KjOwn<{}>", display(T::__typename),),
+        }
+    }
+
+    /// Returns a raw const pointer to the object owned by this [`KjOwn`] if
+    /// any, otherwise the null pointer.
+    #[must_use]
+    pub fn as_ptr(&self) -> *const T {
+        let this = std::ptr::from_ref::<Self>(self).cast::<c_void>();
+        unsafe { T::__get(this) }
+    }
+}
+
+/// Represents a type which can be held in a [`KjOwn`] smart pointer.
+/// # Safety
+/// Cannot be implmented outside of generated workerd-cxx code.
+pub unsafe trait KjOwnTarget {
+    #[doc(hidden)]
+    fn __typename(f: &mut fmt::Formatter) -> fmt::Result;
```

Comment:
is this all to make KjOwn a nice Debug? I think simple `KjOwn(hex_ptr, hex_ptr)` will suffice.